sys/aarch64/rootdev.c

```diff
@@ -45,11 +77,49 @@ static bus_space_t *rootdev_bus_space = &(bus_space_t){
 
 static void rootdev_intr_setup(device_t *dev, unsigned num,
                                intr_handler_t *handler) {
-  /* TODO(pj) */
+  rootdev_t *rd = dev->parent->state;
+
+  assert(num < N_IRQ);
+  intr_event_t *event = &rd->intr_event[num];
+  intr_event_add_handler(event, handler);
 }
 
 static void rootdev_intr_teardown(device_t *dev, intr_handler_t *handler) {
-  /* TODO(pj) */
+  intr_event_remove_handler(handler);
+}
+
+/* Read 32 bit pending register located at va and run handlers. */
+static inline void intr_handle(vaddr_t va, intr_event_t *events) {
+  assert(va != 0);
+  assert(events != NULL);
+
+  uint32_t pending = *(uint32_t *)va;
+
+  for (uint32_t irq = 0; irq < 32; irq++) {
+    if (pending & (1 << irq))
+      intr_event_run_handlers(&events[irq]);
+  }
+}
+
+static void rootdev_intr_handler(device_t *dev, void *arg) {
+  assert(dev != NULL);
+  rootdev_t *rd = dev->state;
+
+  /* Handle local interrupts. */
+  intr_handle(rootdev_local_handle + BCM2836_LOCAL_INTC_IRQPENDINGN(0),
+              &rd->intr_event[LOCAL_OFFSET]);
+
+  /* Handle base interrupts. */
+  intr_handle(rd->arm_base + (BCM2835_INTC_IRQBPENDING & OFFSET_MASK),
```

Comment:
Can You explain why we use OFFSET_MASK here and in the lines below? I can't seem to find a case where it makes a difference i.e. BCM2835_INTC_IRQBPENDING is already a relative address (because BCM2835_INTC_BASE is).

Reply:
Thanks! It's my mistake.
It should be `BCM2835_ARMICU_BASE` + `BCM2835_INTC_*`.
It should make more sense now.